### ISSUE TITLE
SPIKE: show SAML requests are being lost on eauth.va.gov

### ISSUE THREAD

**Issue by ericbuckley** [department-of-veterans-affairs/va.gov-team#13136]
One theory as to low SSOe success rates is that users are not being forwarded from eauth.va.gov to ID.me every time (maybe they are seeing an error screen?).  If we can show this using log results from www.va.gov, eauth.va.gov and ID.me; that would provide enough confidence to the teams that more work needs to be done on the eauth.va.gov servers to support SSOe integration.

Right now we have access to the www.va.gov app logs as well as the eauth logs via splunk.  If we write a script to consume both sources, we can calculate a delta of SAML Requests that fail to get a Response, but do show up in the eauth logs.  This scenario is expected when the user stops the authentication flow because they forgot their password or had problems with the specific credential provider (ID.me, MHV, DSLogon).

If we also can obtain logging records on the ID.me side to see how many SAML Requests were started by users, we could compare that count to see if there is a fall off after eauth, but before ID.me.

For example; lets say that for time period X we have the following results for ID.me LOA1 SAML Requests
- 100 SAML Requests generated by www.va.gov
- 95 SAML Requests received by eauth.va.gov
- 75 SAML Requests received by ID.me

The above would show us that there is a bigger problem between eauth and ID.me than between VA.gov and eauth; in this case more work would need to be put into the integration of those 2 systems to see what the underlying problem may be.

**Comment by ericbuckley** [comment]
Analysis was done using the new "tracker" stat emitted on the SAML Post binding form to show that there are some requests not even being sent out to eauth.  Given the following information, it seems that we are consistently not sending out 1%-2% of the requests we are generating.  Two possible causes for this are; a) JS errors are occurring in the browser and preventing the form from being submitted b) nginx is not correctly forwarding the response from the application server.

We should follow up on the potential JS errors with a new PR to integrate sentry tracking in the form.   As for the nginx issues, more analysis should be done to see if the web server is creating errors at a higher rate than v0 requests.

```
dslogon: 09/18/2020 16:00 EDT -> 09/21/2020 16:00 EDT
{
  "SAML Responses": {
    "v0 [resp/req]": "78.43 [54550/69552]",
    "v1 [resp/req]": "69.57 [2817/4049]",
    "v1 [resp/tracked]": "71.73 [2817/3927]"
  }
}
mhv: 09/18/2020 16:00 EDT -> 09/21/2020 16:00 EDT
{
  "SAML Responses": {
    "v0 [resp/req]": "76.56 [29922/39082]",
    "v1 [resp/req]": "66.67 [1582/2373]",
    "v1 [resp/tracked]": "68.84 [1582/2298]"
  }
}
idme: 09/18/2020 16:00 EDT -> 09/21/2020 16:00 EDT
{
  "SAML Responses": {
    "v0 [resp/req]": "90.65 [53421/58933]",
    "v1 [resp/req]": "82.27 [2784/3384]",
    "v1 [resp/tracked]": "83.35 [2784/3340]"
  }
}
```

**Comment by ericbuckley** [comment]
I ran a script yesterday to look analyze the nginx access logs and see what percentage of authentication requests result in a non success response (302 for v0 and 200 for v1).  The thought was that if our v1 requests had a higher percentage of web server errors than v0, maybe that difference could explain our decrease in success rates.  The results over a 72 hour window did show a higher rate of errors, but nothing that could explain the 6%-7% gap we see with response rates.

```
dslogon: 09/21/2020 13:00 EDT -> 09/24/2020 13:00 EDT
{
  "SAML nginx Errors": {
    "v0 [err/req]": "0.10 [130/134271]",
    "v1 [err/req]": "0.41 [32/7876]"
  }
}
mhv: 09/21/2020 13:00 EDT -> 09/24/2020 13:00 EDT
{
  "SAML nginx Errors": {
    "v0 [err/req]": "0.10 [68/64794]",
    "v1 [err/req]": "0.43 [17/3983]"
  }
}
idme: 09/21/2020 13:00 EDT -> 09/24/2020 13:00 EDT
{
  "SAML nginx Errors": {
    "v0 [err/req]": "0.10 [63/62515]",
    "v1 [err/req]": "0.82 [31/3797]"
  }
}
```

**Comment by joeniquette** [comment]
Completed [writeup](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/teams/vsp/teams/identity/Cookie_Tracer) of VA.gov data analysis. Waiting on id.me and IAM to produce data for comparison.

Next steps, which likely can be a new ticket, are:

1. Investigate the requests that are missing a response.
2. Determine a solution that will allow for the determination of a request tied to a user "eventually being able to successfully login"
3. Create logging logic which allows for automated request and response matching
    1. run statistical analysis of missing responses (such as browser type, etc.)

Data of 0900 ET 4/21/21 - 0900 ET 4/22/21:
![Screen Shot 2021-05-05 at 11.18.23 AM.png](https://images.zenhubusercontent.com/5f6a5e454ede530876cfc425/1b50f7f9-4687-44bd-b2cf-774661555150)

**Comment by joeniquette** [comment]
Received logs from ID.me on 5/12 at noon ET. Picking this ticket back up.

**Comment by joeniquette** [comment]
As of 5/24/21 waiting on query from Splunk to complete. Need to move this to sprint 53.

**Comment by joeniquette** [comment]
Update: 5/26/21 - Resorting to pulling one hour blocks of data from splunk manually. IAM team has not responded to why certain log events are missing from Splunk.

**Comment by joeniquette** [comment]
update: 6/7/21 - received 24 hours of data from IAM with update query params. Plan to complete analysis by EOD Wednesday.

**Comment by joeniquette** [comment]
Analysis and presentation completed: https://docs.google.com/presentation/d/16Uk1nhHmYnyD3Z5BF6p6x6Q44OtIP6tYDtK9hLnvhCw/edit#slide=id.gb6da9f5cd1_0_8

Once version of slide deck is approved, will post ppt and pdf here.

**Comment by joeniquette** [comment]
[https://app.zenhub.com/files/133843125/c1dd7410-9562-4eda-8898-79c54e114e69/download](https://app.zenhub.com/files/133843125/c1dd7410-9562-4eda-8898-79c54e114e69/download)

**Comment by Bootboiler** [comment]
I am a VA user, and I use Safari.  I have have 3 of the possible logins, the DS authentication is not working and has not for months so I don't have that one.  Sometimes I get in, mostly I am unable to access the VA Health app.  I get bad SAML authentication, after I have entered my User name, and correct PW, and then get a code, which I enter.  Sometimes I actually get to the VA site, but I then click on Messages and it goes bad.  

This is the usual message on the id.me, login or the my healthevet.  
![image](https://user-images.githubusercontent.com/104215895/164745228-605c0eb2-73e4-4a59-b936-9dafae5ece7d.png)

If I click log into my healthevet only button, all works fine.  Hope this helps.

![image](https://user-images.githubusercontent.com/104215895/164746782-95aef23c-263b-4bad-b38d-42e13e3fa5fe.png)
